### ISSUE TITLE
Filter amplication artifacts

### ISSUE THREAD

**Issue by jts** [andersen-lab/ivar#75]
**Is your feature request related to a problem? Please describe.**

In multiplexed PCR protocols, aberrant amplicons can be generated which cross the expected amplicon boundaries (for example a fragment may start in amplicon 5 and extend all the way to amplicon 7). In ligation-based sequencing protocols these are easy to identify based on the primers on either end of the paired reads. With Nextera-based protocols it is more difficult but populations of paired-end fragments that begin in one amplicon and end in another are fairly straightforward to identify. In my experience these artifacts can lower the allele frequency of SNPs in primer binding sites, occasionally by as much as 10-15%.

**Describe the solution you'd like**

I wrote a simple python prototype to filter these, the core of it is here:

```
    # make an interval tree containing the amplicons
    tree = IntervalTree()
    for a in amplicons:
        tree.addi(amplicons[a].start, amplicons[a].end, a)

    bam_file = pysam.AlignmentFile(args.bam, "rb")
    bam_out = pysam.AlignmentFile("-", "w", template=bam_file)

    for alignment in bam_file.fetch(until_eof=True):
        fragment_start = alignment.reference_start
        fragment_end = 0
        tlen = alignment.template_length
        if tlen < 0:
            fragment_start = alignment.reference_end
        fragment_end = fragment_start + tlen

        start_intervals = tree.at(fragment_start)
        end_intervals = tree.at(fragment_end)

        # make a set of amplicon IDs that are compatible with the fragment start/end
        start_id_set = set([x[2] for x in start_intervals])
        end_id_set = set([x[2] for x in end_intervals])
        id_intersection = start_id_set.intersection(end_id_set)

        if len(id_intersection) == 1:
            bam_out.write(alignment)
```

This creates an interval tree containing from the coordinates of each amplicon, then looks to see if the read pair is compatible with exactly one set of coordinates to ensure it does not cross the amplicon boundary. I'm considering adding this python script into our fork of the COG-UK nextflow (cc @m-bull) but it would be cleaner if this feature could be added to `ivar trim` (probably best to put it behind a flag that is defaulted to off) to avoid an extra pass through the BAM file.

Thanks!

**Comment by gkarthik** [comment]
Hello Jared, 10-15% is quite a bit! We've been thinking about this a little bit on our end too. The only issue with the approach you show in the pseudo code is that V3 artic scheme contains multiple primers for the same region (`nCoV-2019_44_LEFT` and `nCoV-2019_44_LEFT_alt3` for example). With the NextEra step, it could be possible for one read to start and end at 2 amplicons. To refine this further, we could look for amplicons that completely enclose a read and if that is equal to 0 we can exclude the read (using a flag) otherwise we keep it. Do you think that makes sense?

**Comment by gkarthik** [comment]
Yes, the case in which the PE read starts and ends within the overlapping regions of two amplicons whose primer positions differ by <10 bases. In those cases, just identifying the number of amplicons that "contain" the read will do the trick. If its 0 that means the read spans more than 1 amplicon and can be excluded and if its >0 it can be retained. We've started work on this and will keep things updated here. 

PS: Thank you for raising this issue! This would be a very useful feature to add.

**Comment by jts** [comment]
Ok, in that case I agree that the test should be for a non-empty intersection.

**Comment by gkarthik** [comment]
Fixed by 7027563fd75581c78dabc6040ebffdee2b24abe6

**Comment by jts** [comment]
Thank you @AlaaALatif and @gkarthik for these changes!
